### PR TITLE
ci(docker): use Depot for multiarch images

### LINKED ISSUE #9043: build SDK release image with Depot
## What is the Problem Being Solved?

The [Build release Docker Images workflow](https://github.com/schnetzlerjoe/agoric-sdk/blob/master/.github/workflows/docker.yml) takes just [<8min for amd](https://github.com/Agoric/agoric-sdk/actions/runs/8150759786/job/22277544382) but  [~2h for arm](https://github.com/Agoric/agoric-sdk/actions/runs/8150759786/job/22277544469).

## Description of the Design

Adopt Depot.dev like we did in https://github.com/Agoric/agoric-3-proposals/pull/88

## Security Considerations
n/a
## Scaling Considerations
n/a

## Test Plan
TBD
## Upgrade Considerations
n/a

### PR DESCRIPTION
**Pull request by michaelfig**
closes: #9043

## Description
Update `docker.yml` image publishing to use [Depot](https://depot.dev) for efficient multiarch builds.

Before (~2h):
image

After (~15m):
image

Also, stop building the `agoric/agoric-sdk:ibc-alpha` tag since it is no longer relevant (only was needed by an IBC test long, long ago).

### Security Considerations

Introduces CI dependency on Depot.  `agoric-3-proposals` already uses this successfully, so it is not new to our organisation.

### Scaling Considerations

Improves scalability of our CI system (2h Docker multiarch build times are reduced to 15m).

### Documentation Considerations

n/a

### Testing Considerations

Manually launched Docker builds show that the new CI works correctly.

### Upgrade Considerations

n/a

### CODE REVIEW

File: .github/workflows/docker.yml [hunk 1]
```diff
@@ -168,67 +110,148 @@ jobs:
...
+
       - name: Save SDK_TAG
         run: echo "SDK_TAG=${{ needs.snapshot.outputs.tag }}" >> $GITHUB_ENV
+
+      - name: Override some Docker tags we don't want
+        run: |
+          IMAGE="$REGISTRY/agoric/agoric-sdk"
+          sed -i.bak -e "s%$IMAGE:unreleased%$IMAGE:$SDK_TAG%g" Dockerfile
+          sed -i.bak -e "s%$REGISTRY/agoric/agoric-3-proposals:%$IMAGE:%g" docker-bake.hcl
+        working-directory: a3p-integration
```
Comment (mhofman) [inline .github/workflows/docker.yml]: This looks like voodoo to me

**Review by mhofman — COMMENTED** [review]
This one is flying a bit above my head. One thing however, I don't see an equivalent PR in `agoric-3-proposals` for the patch here.

File: a3p-integration/.yarn/patches/@agoric-synthetic-chain-npm-0.1.1-d26a740f05.patch [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+--- a/dist/cli/cli.js
++++ b/dist/cli/cli.js
+@@ -14,7 +14,9 @@ import path from "node:path";
+ var defaultConfig = {
+   // Tag of the agoric-3 image containing all passed proposals
+   // Must match the Bake file and CI config
+-  fromTag: "latest"
++  fromTag: "latest",
++  repositoryColon: "ghcr.io/agoric/agoric-3-proposals:",
++  sdkRepositoryColon: "ghcr.io/agoric/agoric-sdk:",
```
Comment (turadg) [inline a3p-integration/.yarn/patches/@agoric-synthetic-chain-npm-0.1.1-d26a740f05.patch]: where is this value changed? 

or `repositoryColon` for that matter?

File: a3p-integration/.yarn/patches/@agoric-synthetic-chain-npm-0.1.1-d26a740f05.patch [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+index c59a54224c101648c5507f451bfe0b2cb7f420a5..24ef27a882805df4a969a96ab28dad15b8ee5479 100755
+--- a/dist/cli/cli.js
++++ b/dist/cli/cli.js
+@@ -14,7 +14,9 @@ import path from "node:path";
+ var defaultConfig = {
+   // Tag of the agoric-3 image containing all passed proposals
+   // Must match the Bake file and CI config
+-  fromTag: "latest"
++  fromTag: "latest",
++  repositoryColon: "ghcr.io/agoric/agoric-3-proposals:",
```
Comment (turadg) [inline a3p-integration/.yarn/patches/@agoric-synthetic-chain-npm-0.1.1-d26a740f05.patch]: why is the colon necessary to include?

Comment (turadg) [reply a3p-integration/.yarn/patches/@agoric-synthetic-chain-npm-0.1.1-d26a740f05.patch]: given there are multiple repositories in play, both git and Docker, please give this a more descriptive name. assuming we can drop the colon,

```suggestion
+  a3pDockerRepository: "ghcr.io/agoric/agoric-3-proposals",
```

File: a3p-integration/.yarn/patches/@agoric-synthetic-chain-npm-0.1.1-d26a740f05.patch [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+-    name: `${repository}:${target}`,
++    name: `${buildConfig.repositoryColon}${target}`,
+     target
+   };
+ }
+@@ -143,7 +145,7 @@ RUN /usr/src/upgrade-test-scripts/run_prepare_zero.sh
+   RESUME(fromTag) {
+     return `
+ ## RESUME
+-FROM ghcr.io/agoric/agoric-3-proposals:${fromTag} as use-${fromTag}
```
Comment (turadg) [inline a3p-integration/.yarn/patches/@agoric-synthetic-chain-npm-0.1.1-d26a740f05.patch]: why are these Docker vars necessary? the Dockerfile is generated by the CI job. Couldn't it just produce the desired file contents?

**Review by turadg — CHANGES_REQUESTED** [review]
So far this seems overly complex to me. Let's figure out how to simplify.

File: a3p-integration/.yarn/patches/@agoric-synthetic-chain-npm-0.1.1-d26a740f05.patch (hunk 4 above)
Comment (michaelfig) [reply a3p-integration/.yarn/patches/@agoric-synthetic-chain-npm-0.1.1-d26a740f05.patch]: The `buildConfig` is part of the `package.json`.  Would you prefer that CI edited that before prepare-build instead of creating a `docker-bake.override.json` to modify the Dockerfile's behaviour at build time?

Comment (turadg) [reply a3p-integration/.yarn/patches/@agoric-synthetic-chain-npm-0.1.1-d26a740f05.patch]: yes, very much

Comment (michaelfig) [reply a3p-integration/.yarn/patches/@agoric-synthetic-chain-npm-0.1.1-d26a740f05.patch]: I've removed any dependencies on a synthetic-chain patch, and reduced the postprocessing of the `prepare-build`-generated files to a single `sed`.  Let's talk about the best way to make that change robust.

File: a3p-integration/package.json [hunk 5]
```diff
@@ -12,7 +12,7 @@
     "doctor": "yarn synthetic-chain doctor"
   },
   "dependencies": {
-    "@agoric/synthetic-chain": "patch:@agoric/synthetic-chain@npm%3A0.1.0#~/.yarn/patches/@agoric-synthetic-chain-npm-0.1.0-148de716a6.patch",
+    "@agoric/synthetic-chain": "^0.1.1",
```
Comment (turadg) [inline a3p-integration/package.json]: might as well bump to 0.3.0

File: .github/workflows/docker.yml [hunk 6]
```diff
@@ -168,67 +110,159 @@ jobs:
...
+        id: meta
+        uses: docker/metadata-action@v5
         with:
-          file: packages/deployment/Dockerfile.ibc-alpha
-          context: packages/deployment/docker
-          platforms: linux/amd64,linux/arm64/v8
+          images: ${{ env.REGISTRY }}/agoric/agoric-sdk
+          tags: ${{ fromJSON(steps.prefix-tags.outputs.json-tags) }}
+
+      - name: Publish "use-upgrade-next" image
```
Comment (turadg) [inline .github/workflows/docker.yml]: mind separating "ci(docker): use Depot multiarch, publish a3p-use-upgrade-next" into separate commits?

even better, separate PRs. One that closes https://github.com/Agoric/agoric-sdk/issues/9043 and another that does this publishing. There's more design to work through on that. I'd really like to see an issue which details the problem to be solved.

**Comment by michaelfig** [comment]
I extracted the controversial "publish `a3p-use-upgrade-next`" code into #10083.

File: .github/workflows/docker.yml [hunk 7]
```diff
@@ -33,109 +34,50 @@ jobs:
...
+        uses: docker/login-action@v3
+        # see https://docs.github.com/en/actions/publishing-packages/publishing-docker-images
+        with:
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+          registry: ${{ env.REGISTRY }}
       - name: Save BUILD_TAG
         run: |
-          ARCH=$(echo '${{ matrix.platform }}' | tr / _)
-          echo "BUILD_TAG=${{ needs.snapshot.outputs.tag }}-$ARCH" >> $GITHUB_ENV
```
Comment (mhofman) [inline .github/workflows/docker.yml:52]: Are we still generating `-$ARCH` tags? I have used those in the past, but mostly because the arm64 and thus combined image took forever to publish, so possibly this is no longer necessary.

**Review by mhofman — COMMENTED** [review]
I don't feel knowledgeable enough with the depot stuff to approve, but from what I'm seeing this look fine to me.

File: .github/workflows/docker.yml (hunk 7 above)
Comment (michaelfig) [reply .github/workflows/docker.yml:52]: You're right; building and publishing the combined image now takes about the same time as the `-linux_amd64` tag used to take.  So, no more `-$ARCH` tags.

File: .github/workflows/docker.yml [hunk 8]
```diff
@@ -251,30 +174,28 @@ jobs:
       - name: Prefix tags
         id: prefix-tags
         run: |
-          IMAGE="$REGISTRY/agoric/cosmic-swingset-solo"
+          PREFIX="$REGISTRY/agoric/cosmic-swingset-solo:"
           for TAG in ${{ needs.docker-sdk.outputs.tags }}; do
-            PREFIXED="$PREFIXED$sep$IMAGE:$TAG"
+            PREFIXED="$PREFIXED$sep$PREFIX$TAG"
```
Comment (turadg) [inline .github/workflows/docker.yml]: why this change? the previous was more clear. please revert or justify.

**Review by turadg — APPROVED** [review]
Nice improvement!

LGTM save for moving `:` into the variable. I trust you'll address before merging

File: .github/workflows/docker.yml (hunk 8 above)
Comment (michaelfig) [reply .github/workflows/docker.yml]: Okay.
